lib/course/dsl/routing.rb

```diff
@@ -5,24 +5,48 @@ def self.included(base)
         base.extend(ClassMethods)
       end
 
+      # rubocop:disable Metrics/AbcSize
       def call(env)
-        return self.class.find_get_endpoint(env).call if handle_get_request?(env)
-        return self.class.find_post_endpoint(env).call if handle_post_request?(env)
+        result = get?(env) && handle_request?(env, self.class.list_of_get)
+        body = self.class.find_endpoint(self.class.list_of_get, result.join("/")).call if result
+        result = post?(env) && handle_request?(env, self.class.list_of_post)
+        body = self.class.find_endpoint(self.class.list_of_post, result.join("/")).call if result
+        return [self.class.status_or_default, self.class.headers_or_default, [body || ""]] unless body.nil?
+
         return self.class.find_rack_app(env).new.call(env) if handle_other_rack_request?(env)
 
         [500, {}, [""]]
       end
+      # rubocop:enable Metrics/AbcSize
+
+      def handle_request?(env, list)
+        arr_request_path = env["PATH_INFO"].split("/")
+        list.each do |path|
+          arr_path = path[0].split("/")
+          next if arr_path.size != arr_request_path.size
 
-      def handle_get_request?(env)
-        return false unless get?(env)
+          find_url = match(arr_path, arr_request_path)
```

Comment:
anyway, good attempt. Know what, let's use an already built solution here: https://github.com/sinatra/mustermann
Hanami and sinatra uses it, and it's quite fast.